### DISCUSSION TITLE
Adding Stride implementation (code only) to existing Maui project?

### DISCUSSION THREAD

**Discussion by jonmdev** [stride3d/stride#2196] [Q&A] [answered]
Let's say you have a Maui project created by File > New > MAUI project in Visual Studio 2022. If you want to add Stride 3D scene to this, is it possible? 

The steps I would imagine would be:

## 1) Add Stride Package

Add Stride to existing Maui project by Package Manager. According to [this](https://stride3d.github.io/stride-community-toolkit/manual/code-only/create-project.html) and [this](https://github.com/stride3d/stride-community-toolkit/tree/main/examples/code-only/Example01_Basic3DScene_VBasic) Stride can do well with code only in .NET by just adding the [Community Toolkit](https://www.nuget.org/packages/Stride.CommunityToolkit) to an existing Visual Studio project. Is this true? 

If you don't need the Editor, can the Community Toolkit alone handle basic tasks like camera, simple FBX models, animations?

## 2) Create & Run Stride Scene

Ideally, perhaps, we can then add FBX's, textures, or other assets to our Resources folder, and then instantiate a 3D scene programmatically somehow from these using the Stride code in the background, with a function run from inside the Maui project to start/stop the Stride scene.

## 3) Map Stride Camera > Maui View 

The output of a camera could then be somehow transferred into a Native view element (TextureView or SurfaceView in Android, SwapChainPanel in WIndows, UIView or NSView in iOS).

This could then be mapped into Maui as a View, for example following [this guide](https://learn.microsoft.com/en-us/dotnet/maui/user-interface/handlers/create?view=net-maui-8.0).

# QUESTION

Is this feasible? Is adding Stride to an existing Maui app (for iOS/Android/Windows support) remotely possible? 

Thanks for any help or ideas.

**Comment by Doprez** [comment, marked as answer]
## 1
Yes! One of the main reasons I love Stride is that its all Nuget packages! even the [game studio editor](https://github.com/Doprez/stride-gamestudio-nuget) which for some reason surprised me.

## 2
This is definitely possible, the only caveat being that the current asset importers from Stride are Windows specific so you may have to use a third party library to import them at runtime on another platform if that is an issue. [this](https://github.com/ykafia/StrideGltf/blob/master/GltfImport/LoadGLTF.cs) is an example of loading GLTF files at runtime so you could use most of the same code but instead use an FBX compatible library.

## 3
Some examples here may help:
WPF - https://gist.github.com/EricEzaM/5797be1f4b28f15e9be53287a02d3d67
Winforms - https://github.com/windperson/-StrideGameNet5WinFormsDemo

You should be able to use the handle of the [GameWindow](https://github.com/stride3d/stride/blob/master/sources/engine/Stride.Games/GameWindow.cs) to render to another UI if you want to see where to look in source you should check [this](https://github.com/stride3d/stride/tree/master/sources/engine/Stride.Games) folder out. Since it is not only Directx I dont think it gives you direct access to the swap chain.

**Reply by Doprez** [reply]
also if you want more examples to look through I maintain a list of Stride specific repos as they come in so if you decide to get a workijng embedded MAUI example I would love to add it https://github.com/Doprez/Awesome-Stride#embedded-stride

**Reply by jonmdev** [reply]
Thanks @Doprez. I will give it my best try. I have only just started with the Stride code only model but it seems very simple and intuitive. Very nice overall.

Unfortunately two Maui integration bugs off the start:

https://github.com/stride3d/stride/issues/2198

https://github.com/stride3d/stride/issues/2197

But if those can be resolved, I will keep chipping away at it and see how it goes.

**Reply by tebjan** [reply]
Also have a look here: https://opencollective.com/stride3d/projects/embed-stride-ui-frameworks

If you manage to get it working and submit a pull request, you will receive a bounty. The final reward will be significantly higher than the amount you currently see.

**Reply by wenleyluk** [reply]
I added stride.community.toolkit (preview 37) to my new created MAUI app (VS2022 built-in template),  I can directly deploy it to Android Simulator (w/o the error occurs in #2198), and it can run the Maui app on Android Simulator successfully!

But after adding stride.community.toolkit.skyboxes (preview 32), it would return below error when deploying to Android Simulator:
1>C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\34.0.52\tools\\**Xamarin.Android.Common.targets(905,2): error MSB3094: "DestinationFiles" refers to 6 item(s), and "SourceFiles" refers to 3 item(s). They must have the same number of items.**
1>Done building project "MauiAppTest02.csproj" -- FAILED.

![error_stride](https://github.com/stride3d/stride/assets/13364379/264471ba-4088-4148-9f93-ec955c58ce7e)

Many thanks!

**Reply by VaclavElias** [reply]
@wenleyluk, I have been working on the fix, I hope I should finalise it today https://discord.com/channels/500285081265635328/1179562410655363132

**Reply by VaclavElias** [reply]
@wenleyluk, please try preview38, just released https://github.com/stride3d/stride-community-toolkit/releases/tag/1.0.0.0-preview.38

**Reply by wenleyluk** [reply]
@VaclavElias ,Thank you for your hard working.

But the error on building for Android is still exits after updated to both Stride.CommunityToolkit & Stride.CommunityToolkit.Skyboxes with preview 38 !!

Is it caused by the Skyboxes nuget package does NOT support build for Android platform ?
![error_stride_2nd](https://github.com/stride3d/stride/assets/13364379/855554c6-81f9-4a6d-aee0-16f236568451)

**Reply by IXLLEGACYIXL** [reply]
> @VaclavElias ,Thank you for your hard working.
> 
> But the error on building for Android is still exits after updated to both Stride.CommunityToolkit & Stride.CommunityToolkit.Skyboxes with preview 38 !!
> 
> Is it caused by the Skyboxes nuget package does NOT support build for Android platform ?
> ![error_stride_2nd](https://github.com/stride3d/stride/assets/13364379/855554c6-81f9-4a6d-aee0-16f236568451)

the skybox itself is the Problem

**Reply by IXLLEGACYIXL** [reply]
there have been multiple reports on #android  in disocrd that the skybox is broken for mobile https://discord.com/channels/500285081265635328/1152588396288544819/1200886532375060592

**Comment by VaclavElias** [comment]
> @VaclavElias ,Thank you for your hard working.
> 
> But the error on building for Android is still exits after updated to both Stride.CommunityToolkit & Stride.CommunityToolkit.Skyboxes with preview 38 !!
> 
> Is it caused by the Skyboxes nuget package does NOT support build for Android platform ? ![error_stride_2nd](https://private-user-images.githubusercontent.com/13364379/320761570-855554c6-81f9-4a6d-aee0-16f236568451.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI2NDg5NzMsIm5iZiI6MTcxMjY0ODY3MywicGF0aCI6Ii8xMzM2NDM3OS8zMjA3NjE1NzAtODU1NTU0YzYtODFmOS00YTZkLWFlZTAtMTZmMjM2NTY4NDUxLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA5VDA3NDQzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgwOGU1ZTYzMDM3MWNjYWVkNDk4NmIyYTk0YTY4NDg1ZjMyNTkzN2ZiZmFmY2ZmOGY3MGEyMzA1MDI5MDQ4NDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.hIt514WGWQnBYALz4dpF2ZsnHzhizB968B5IxoCxKhE)

Well, you don't need to use the Skybox Nuget. It is optional, `Stride.CommunityToolkit` NuGet is not referencing it.

If you would like, you can fork also Stride.CommunityToolkit and Stride.CommunityToolkit.Skyboxes and see if you can fix it, maybe `Stride.CommunityToolkit.Skyboxes` needs more set up?

But if you remove `Stride.CommunityToolkit.Skyboxes`, at least, you will see if your build is still working?
